hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAssociateToAccountHandler.java

```diff
@@ -42,22 +66,228 @@ public TokenAssociateToAccountHandler() {
+            // token relation's previous and next token IDs.
+
+            // Compute the previous and next token IDs. Unfortunately `TokenRelation` doesn't
+            // allow for null values, so a value of '0' will have to indicate a null pointer to
+            // the previous or next token (since no token number 0 can exist)
+            long prevTokenId = 0;
+            long nextTokenId = 0;
+            if (i - 1 >= 0) { // if there is a previous token
+                prevTokenId = Optional.ofNullable(tokens.get(i - 1))
+                        .map(Token::tokenNumber)
+                        .orElse(0L);
+            }
+            if (i + 1 < tokens.size()) { // if there is a next token
+                nextTokenId = Optional.ofNullable(tokens.get(i + 1))
+                        .map(Token::tokenNumber)
+                        .orElse(0L);
+            }
+
+            // Create the new token relation
+            final var isFrozen = token.hasFreezeKey() && token.accountsFrozenByDefault();
+            final var kycGranted = !token.hasKycKey();
+            final var newTokenRel = new TokenRelation(
+                    token.tokenNumber(),
+                    account.accountNumber(),
+                    0,
+                    isFrozen,
+                    kycGranted,
+                    false,
+                    false,
+                    prevTokenId,
+                    nextTokenId);
+            newTokenRels.add(newTokenRel);
+        }
+
+        // Now all the NEW token relations are linked together, but they are not yet linked to the account. First,
+        // compute where the account's current head token number should go in the linked list of tokens
+        final var currentHeadTokenNum = account.headTokenNumber();
+        // NOTE: if currentHeadTokenNum is less than 1, it means the account isn't associated with any tokens yet, so
+        // we'll just set the head to the first token, i.e. the first token ID list from the transaction (since the new
+        // tokenRels are all linked, and in the order of the token IDs as they appeared in the original list)
+        if (isValidTokenNum(currentHeadTokenNum)) {
+            // The account is already associated with some tokens, so we need to insert the new
+            // tokenRels at the beginning of the list of existing token numbers first. We start by
+            // retrieving the token rel object with the currentHeadTokenNum at the head of the
+            // account
+            final var headTokenRel = tokenRelStore
+                    .get(
+                            AccountID.newBuilder()
+                                    .accountNum(account.accountNumber())
+                                    .build(),
+                            TokenID.newBuilder().tokenNum(currentHeadTokenNum).build())
+                    .orElse(null);
+            validateTrue(headTokenRel != null, TOKEN_NOT_ASSOCIATED_TO_ACCOUNT);
```

Comment:
I think `validateTrue()` is meant to enforce business logic checks---i.e., to catch user errors.

But if an account has a head token number that _it isn't actually associated to_, that means we have violated an internal system invariant; i.e., this is a bug, not user error.

So instead here we'd want to just log at `ERROR` to surface this problem ASAP; and then preferably work around it and continue handling this transaction.

If it's too hard to work around silently, I'd throw an `IllegalStateException` to rollback the transaction and end with `FAIL_INVALID`.

Reply:
How would we silently work around it? If the token rel is not there, an NPE is thrown on line 179

Reply:
We'll only update the existing linked list relationships if the head token number relation is found. Otherwise we'll log an error and continue with the transaction.

Reply:
Done